### PR TITLE
les 9

### PR DESCRIPTION
**Pull request by andstartut**
Stage 1

### CODE REVIEW

File: Lesson9/src/enums/ProductType.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+public enum ProductType {
+    SEDAN("sedan"),
+    WAGON("wagon"),
+    COUPE("coupe"),
+    PICKUP("pickup"),
+    SUV("SUV"),
+    VAN("VAN"),
+    SPORTCAR("sportcar"),
+    RACE("race");
+    String type;
```
Comment (GlebLitvinov) [inline Lesson9/src/enums/ProductType.java:12]: make it private

File: Lesson9/src/shop/Shop.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.*;
+
+public class Shop implements IShop {
+    List<Product> productList = new ArrayList<>();
+
+    @Override
+    public void addProduct(Product product) throws SearchProductException {
+        if (ShopUtil.isTheProductPresent(productList, product.getId())) {
+            throw new SearchProductException(product.getName() + " is already in the list");
+        }
```
Comment (GlebLitvinov) [inline Lesson9/src/shop/Shop.java:16]: good

File: Lesson9/src/shop/Shop.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    public void addProduct(List<Product> newList) throws SearchProductException {
+        for (Product product : productList) {
+            for (Product newProduct : newList) {
+                if (newProduct.getId() == (product.getId())) {
+                    throw new SearchProductException(newProduct.getName() + " is already in the list");
+                }
+            }
+        }
+        productList.addAll(newList);
```
Comment (GlebLitvinov) [inline Lesson9/src/shop/Shop.java:28]: I'd replace it with
for (Product product: newList) {
addProduct(product);
}

in this case you will be 100% sure that all products are added In the same way

File: Lesson9/src/shop/Shop.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    @Override
+    public boolean deleteProduct(int id) throws SearchProductException {
+        if (ShopUtil.searchProductByID(productList, id) == null) {
+            throw new SearchProductException("The Product with this ID:" + id + " was not found");
+        } else {
+            Product product = ShopUtil.searchProductByID(productList, id);
+            productList.remove(product);
+            return true;
+        }
```
Comment (GlebLitvinov) [inline Lesson9/src/shop/Shop.java:49]: refactor it to look like
 Product product = ShopUtil.searchProductByID(productList, id);
if (product == null) {
            throw new SearchProductException("The Product with this ID:" + id + " was not found");
        } else {
            productList.remove(product);
            return true;
        }
        
        so you don't search for the same id twice

File: Lesson9/src/shop/ShopUtil.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+    public static boolean isTheProductPresent(List<Product> productList, int id) {
+        boolean hasPresent = false;
+        for (Product product : productList) {
+            if (product.getId() == id) {
+                hasPresent = true;
+                break;
+            }
+        }
+        return hasPresent;
```
Comment (GlebLitvinov) [inline Lesson9/src/shop/ShopUtil.java:27]: can be refactored to
return searchProductByID(productList, id) != null;

**Review by GlebLitvinov — APPROVED** [review]
overall is very good (don't forget that you should not merge this PR until the last lesson

File: Lesson9/src/shop/Shop.java (hunk 3 above)
Comment (andstartut) [reply Lesson9/src/shop/Shop.java:28]: If I replace it in the second for, I will not check condition with another products from the first for.

Comment (andstartut) [reply Lesson9/src/shop/Shop.java:28]: Ha! I know decision! I need to swap both for

**Pull request merged**